pkg/kubelet/kuberuntime/kuberuntime_sandbox.go

```diff
@@ -0,0 +1,172 @@
+	podSandboxConfig := &runtimeApi.PodSandboxConfig{
+		Name:        &sandboxName,
+		Labels:      newPodLabels(pod),
+		Annotations: newPodAnnotations(pod),
+	}
+
+	if !kubecontainer.IsHostNetworkPod(pod) {
+		dnsServers, dnsSearches, err := m.runtimeHelper.GetClusterDNS(pod)
+		if err != nil {
+			return nil, err
+		}
+		podSandboxConfig.DnsOptions = &runtimeApi.DNSOption{
+			Servers:  dnsServers,
+			Searches: dnsSearches,
+		}
+		// TODO: Add domain support in new runtime interface
+		hostname, _, err := m.runtimeHelper.GeneratePodHostNameAndDomain(pod)
+		if err != nil {
+			return nil, err
+		}
+		podSandboxConfig.Hostname = &hostname
+	}
+
+	cgroupParent := ""
+	portMappings := []*runtimeApi.PortMapping{}
+	for _, c := range pod.Spec.Containers {
+		opts, err := m.runtimeHelper.GenerateRunContainerOptions(pod, &c, podIP)
+		if err != nil {
+			return nil, err
+		}
+
+		for idx := range opts.PortMappings {
+			port := opts.PortMappings[idx]
+			hostPort := int32(port.HostPort)
+			containerPort := int32(port.ContainerPort)
+			protocol := toRuntimeProtocol(port.Protocol)
+			portMappings = append(portMappings, &runtimeApi.PortMapping{
+				HostIp:        &port.HostIP,
+				HostPort:      &hostPort,
+				ContainerPort: &containerPort,
+				Protocol:      &protocol,
+				Name:          &port.Name,
+			})
+		}
+
+		// TODO: refactor kubelet to get cgroup parent for pod instead of containers
+		cgroupParent = opts.CgroupParent
+	}
+
+	if len(portMappings) > 0 {
+		podSandboxConfig.PortMappings = portMappings
+	}
+
+	podSandboxConfig.Resources = generateSandboxResources(pod)
+	podSandboxConfig.Linux = generatePodSandboxLinuxConfig(pod, cgroupParent)
+
+	return podSandboxConfig, nil
+}
+
+// generateSandboxResources generates runtimeApi.PodSandboxResources from api.Pod
+func generateSandboxResources(pod *api.Pod) *runtimeApi.PodSandboxResources {
+	podResourceLimits := make(map[api.ResourceName]int64)
+	podResourceRequests := make(map[api.ResourceName]int64)
+	for _, c := range pod.Spec.Containers {
+		for name, limit := range c.Resources.Limits {
+			if l, ok := podResourceLimits[name]; ok {
+				podResourceLimits[name] = l + limit.MilliValue()
+			} else {
+				podResourceLimits[name] = limit.MilliValue()
+			}
+		}
+		for name, req := range c.Resources.Requests {
+			if l, ok := podResourceRequests[name]; ok {
+				podResourceRequests[name] = l + req.MilliValue()
+			} else {
+				podResourceRequests[name] = req.MilliValue()
+			}
+		}
+	}
+
+	var podCPULimit, podMemoryLimit, podCPURequest, podMemoryRequest int64
+	for k, v := range podResourceLimits {
+		switch k {
+		case api.ResourceCPU:
+			podCPULimit = v
+		case api.ResourceMemory:
+			podMemoryLimit = v
+		}
+	}
+	for k, v := range podResourceRequests {
+		switch k {
+		case api.ResourceCPU:
+			podCPURequest = v
+		case api.ResourceMemory:
+			podMemoryRequest = v
+		}
+	}
+
+	cpuResource := &runtimeApi.ResourceRequirements{}
+	cpulimit := float64(podCPULimit) / 1000
```

Comment:
Potentially if non of the containers specify the cpu limits, then podCPULimit here is 0. Same thing for other resources.